Fix: do not autofix octal escape sequence (fixes #10031)

malcolmsgroves opened the pull request:
<!--
    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).
-->

**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[X] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->
https://github.com/eslint/eslint/issues/10031
<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/developer-guide/contributing/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

**What changes did you make? (Give an overview)**
Added a check for octal escape sequences in binary expressions. The```prefer-template``` rule does not attempt to fix binary expressions that contain octal escape sequences, and returns a ```null``` fixer object if this is the case.

Also added a test that ensures that binary expressions with octal escape sequences are not auto fixed.

**Is there anything you'd like reviewers to focus on?**

malcolmsgroves commented:
@platinumazure could you take a look at this when you have a moment? Thanks!

not-an-aardvark commented on lib/rules/prefer-template.js:
```diff
@@ -38,6 +38,45 @@ function getTopConcatBinaryExpression(node) {
+        return false;
+    }
+
+    const match = node.raw.match(/^([^\\]|\\[^0-7])*\\([0-3][0-7]{1,2}|[4-7][0-7]|[0-7])/);
```
I think the second part of this regular expression could be simplified from this:

```
([0-3][0-7]{1,2}|[4-7][0-7]|[0-7])
```

to this:

```
([0-7]{1,3})
```

...since it seems like the `[0-7]` character class will always match something even if the other cases don't.

not-an-aardvark commented on tests/lib/rules/prefer-template.js:
```diff
@@ -187,6 +187,11 @@ ruleTester.run("prefer-template", rule, {
+        {
+            code: "foo + 'does not autofix octal escape sequence' + '\\033'",
+            output: null,
+            errors
```
The regular expression you're using is fairly complicated. It seems correct as far as I can tell, but there are a lot of non-obvious cases, so I want to make sure we don't accidentally break those cases if we change the logic in the future. Can you add some tests for cases like the following?

* The string contains an unrelated escape before the octal escape: `` `\n other text \033` + foo ``
* The string contains a `\0` escape followed by an octal escape: `` `\0 other text \033` `` (I'm not sure this case is handled correctly right now)
* The string contains an escaped backslash: `` `\\033` + foo `` (this should get autofixed because there is not actually an octal escape).

malcolmsgroves commented:
Thanks for the comments! I should be able to make the changes after the weekend

malcolmsgroves commented:
@not-an-aardvark I added those tests and cleaned that regex up. You were right – it was mishandling ```\0``` null escape sequences, but should be good now. 

@platinumazure do you think I could work up a PR to move the octal matching to ast-utils?

platinumazure commented:
@malcolmsgroves

> do you think I could work up a PR to move the octal matching to ast-utils?

Absolutely; certainly no need to do it in this PR. Might be good to see if other ESLint core rules could use it as well. Thanks!

not-an-aardvark commented:
Thanks for continuing to work on this. Can you add a test for the '\0 other text \033' case (where an octal escape follows a `\0` escape)? From quickly glancing at the code it's not clear to me that this is fixed, since it seems like `isOctalEscapeSequence` simply returns false whenever `\0` appears in the string.

malcolmsgroves commented:
@not-an-aardvark done, sorry I forgot to include the tests in the original changes. ```isOctalEscapeSequence``` seems to work fine

not-an-aardvark commented on tests/lib/rules/prefer-template.js:
```diff
@@ -34,7 +34,8 @@ ruleTester.run("prefer-template", rule, {
         "var foo = `foo` + `bar` + \"hoge\";",
-        "var foo = `foo` +\n    `bar` +\n    \"hoge\";"
+        "var foo = `foo` +\n    `bar` +\n    \"hoge\";",
+        "var foo = `foo` + `\\0`;"
```
I think these "valid" tests aren't working correctly -- the strings in these tests are already template literals, so the rule isn't reporting them because there would be nothing to adjust anyway.

From testing out this branch locally, I think the string `\0` is still preventing an autofix.

malcolmsgroves commented:
@not-an-aardvark should be good now, I changed the regex a bit and fixed those tests – string literals that only contain ```\0``` are now auto fixed
